articles/qos_configurability.md

```diff
@@ -0,0 +1,374 @@
+  Until a better QoS profile for services and actions is designed, this feature won't be exposed.
+- Shall only allow changing QoS at startup time, and not during runtime by default.
+  - Rationale: Allowing runtime reconfiguration of QoS is a complex topic, and purposefully left out of scope in this proposal.
+- Shall have a way of reusing the same QoS profile.
+  - Rationale: Not having a way of setting the same QoS profile to different entities will make configuring entities which profiles must match (or be compatible) error prone.
+
+## Adding a new mechanism vs using parameters
+
+A previous proposal (see PR [#296](https://github.com/ros2/design/pull/296)) suggested using a new file format for reconfiguring QoS policies.
+Some concerns were raised by the community:
+1. It adds a different mechanism to the ones that already exist.
+    Newcomers have to learn yet another thing.
+1. QoS settings are not conceptually different to other parameters, e.g.: publishing rate.
+1. It adds a new mechanism to configure QoS profiles.
+
+Based on that feedback, this PR will explore a design based on parameters.
+
+## Introduction
+
+Current code will not allow QoS reconfigurability if not changed:
+
+```cpp
+node->create_publisher(
+  "chatter",
+  KeepLast{10});
+```
+
+To make reconfigurability easy, only a flag is needed to automatically create the parameters:
+
+```cpp
+node->create_publisher(
+  "chatter",
+  KeepLast{10},
+  QosOverridingOptions{true});  // allow_reconfigurable all qos
+```
+
+That will automatically declare the parameters for reconfiguring the QoS policies that can be overridden in the following way:
+
+```yaml
+/my/ns/node_name:
+  ros__parameters:
+    qos_profiles:  # group for all the qos settings
+      publisher:
+        'my/fully/qualified/topic/name/here': 
+          reliability: reliable
+          history_depth: 100
+          history: keep_last
+```
+
+A callback to accept/reject the configured profile can also be specified:
+
+```cpp
+node->create_publisher(
+  "chatter",
+  KeepLast(10),
+  [] (const QoSProfile & qos) -> bool {
+      return qos.is_reliable();  // only accept reliable communication, we should add getters to rclcpp::QoSProfile class for this.
+  });
+```
+
+The user provided QoS callback will be internally used as a [parameters callback](https://github.com/ros2/rclcpp/blob/3a4ac0ca2093d12035070443692798b0c9f9da3a/rclcpp/include/rclcpp/node_interfaces/node_parameters_interface.hpp#L183), but we provide a more user-friendly interface.
+
+## Analysis of parameter API features and limitations
+
+### Read-only parameters
+
+ROS 2 currently provides read only parameters.
+They can be modified at startup time by providing an override (e.g. `--ros-args -p <param_name> <param_value>`), but they cannot be dynamically changed later.
```

Comment:
> However, it is possible to create nodes post-launch. A node created later will also need to read its parameters and so, technically, the user can specify the parameter overrides on the command-line at startup/launch time, but they are read/modified/overridden post-launch.

I'm not sure I understand.
"startup-time" and "launch time" are vague terms, so I will try to rephrase this.

The idea is that you cannot modify the parameter after it was declared (because it's a *const* param), but you can have parameter overrides that are created at runtime (the same applies to other consts parameters).
Parameter overrides can only be provided at node construction time, so yes you would be able to tear down the node, modify the overrides and launch it again.

As another example, when loading a node in a container the message to load the node can have [parameter overrides](https://github.com/ros2/rcl_interfaces/blob/9b7bf4b30e2f51c09fdac86ba711fd6a0f1a5910/composition_interfaces/srv/LoadNode.srv#L27), so it's possible to specify the [parameter overrides at runtime](https://github.com/ros2/rclcpp/blob/5851eebdda5cc2ed9186a034df0ef75ccfe032d3/rclcpp/include/rclcpp/node_options.hpp#L131).

Reply:
Yes, agreed. Just wanted to be clear about the term "startup-time." Perhaps referring to node construction time specifically instead would avoid any potential misunderstandings (that is, misinterpreting "start-up time" as being "launch time").

Reply:
Clarified in https://github.com/ros2/design/pull/300/commits/f8f27f6b902021b75eebbbf3cf6582ae484c0170.